pkg/oc/cli/image/extract/extract.go

```diff
@@ -0,0 +1,576 @@
+
+		  [<index>] - select the layer at the provided index (zero-indexed)
+		  [<from_index>,<to_index>] - select layers by index, exclusive
+		  [~<prefix>] - select the layer with the matching prefix or return an error
+
+		Negative indices are counted from the end of the list, e.g. [-1] selects the last
+		layer.
+
+		Experimental: This command is under active development and may change without notice.`)
+
+	example = templates.Examples(`
+# Extract the busybox image into the current directory
+%[1]s docker.io/library/busybox:latest
+
+# Extract the busybox image to a temp directory (must exist)
+%[1]s docker.io/library/busybox:latest=/tmp/busybox
+
+# Extract a single file from the image into the current directory
+%[1]s docker.io/library/centos:7=/bin/bash=.
+
+# Extract all .repo files from the image's /etc/yum.repos.d/ folder.
+%[1]s docker.io/library/centos:7=/etc/yum.repos.d/*.repo=.
+
+# Extract the last layer in the image
+%[1]s docker.io/library/centos:7[-1]
+
+# Extract the first three layers of the image
+%[1]s docker.io/library/centos:7[:3]
+
+# Extract the last three layers of the image
+%[1]s docker.io/library/centos:7[-3:]
+`)
+)
+
+type Options struct {
+	Mappings []Mapping
+
+	CreatedAt string
+
+	OnlyFiles bool
+
+	FilterOptions imagemanifest.FilterOptions
+
+	MaxPerRegistry int
+
+	DryRun   bool
+	Insecure bool
+	Force    bool
+
+	genericclioptions.IOStreams
+
+	ImageMetadataCallback func(m *Mapping, dgst digest.Digest, imageConfig *docker10.DockerImageConfig)
+}
+
+func NewOptions(streams genericclioptions.IOStreams) *Options {
+	return &Options{
+		IOStreams:      streams,
+		MaxPerRegistry: 3,
+	}
+}
+
+// New creates a new command
+func New(name string, streams genericclioptions.IOStreams) *cobra.Command {
+	o := NewOptions(streams)
+
+	cmd := &cobra.Command{
+		Use:     "extract",
+		Short:   "Copy files from an image to the filesystem",
+		Long:    desc,
+		Example: fmt.Sprintf(example, name),
+		Run: func(c *cobra.Command, args []string) {
+			kcmdutil.CheckErr(o.Complete(c, args))
+			kcmdutil.CheckErr(o.Run())
+		},
+	}
+
+	flag := cmd.Flags()
+	o.FilterOptions.Bind(flag)
+
+	flag.BoolVar(&o.DryRun, "dry-run", o.DryRun, "Print the actions that would be taken and exit without writing to the destination.")
+	flag.BoolVar(&o.Insecure, "insecure", o.Insecure, "Allow push and pull operations to registries to be made over HTTP")
+
+	flag.StringVar(&o.CreatedAt, "created-at", o.CreatedAt, "The creation date for this image, in RFC3339 format or milliseconds from the Unix epoch.")
+
+	flag.BoolVar(&o.OnlyFiles, "only-files", o.OnlyFiles, "Only extract regular files and directories from the image.")
+
+	flag.BoolVar(&o.Force, "force", o.Force, "If set, the command will attempt to upload all layers instead of skipping those that are already uploaded.")
+	flag.IntVar(&o.MaxPerRegistry, "max-per-registry", o.MaxPerRegistry, "Number of concurrent requests allowed per registry.")
```

Comment:
when would there be more than one registry involved?  (carry-over from image mirror?)

Reply:
You can extract two images from different registries

Reply:
is it necessary that we support that?  I think allowing multiple images to extract from, in a single commandline, is going to frustrate our ability to support args in a sane way.  Why not just support a single image to extract from, per invocation?

Among other things, this would make it easier to support multiple paths to extract (e.g. "--extract /tmp --extract /etc").